Yaml decode error

```
Running golang arch lint...
docker run --rm  -v ***:/app \
        fe3dback/go-arch-lint:release-v1.4.0 check --project-path /app
panic: runtime error: invalid memory address or nil pointer dereference
--
goroutine 1 [running]:
runtime/debug.Stack(0xa093038, 0x8251680, 0x84065d8)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x83
github.com/fe3dback/go-arch-lint/cmd.Execute.func1(0xa0bc060, 0xa093fc0)
        /tmp/builder/cmd/cmd_root.go:59 +0xf8
panic(0x8251680, 0x84065d8)
        /usr/local/go/src/runtime/panic.go:969 +0x122
github.com/goccy/go-yaml.(*Decoder).createDecodedNewValue(0xa0c62a0, 0x82d0800, 0x82455a0, 0x0, 0x0, 0x1, 0x8238600, 0xa0afbe0, 0x97, 0x0)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:650 +0x1d
github.com/goccy/go-yaml.(*Decoder).decodeSlice(0xa0c62a0, 0x8238600, 0xa0afba0, 0x197, 0x82d0500, 0xa1040e0, 0x81d4676, 0x8249700)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:1028 +0x245
github.com/goccy/go-yaml.(*Decoder).decodeValue(0xa0c62a0, 0x8238600, 0xa0afba0, 0x197, 0x82d0500, 0xa1040e0, 0x12, 0x0)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:557 +0x1bf3
github.com/goccy/go-yaml.(*Decoder).createDecodedNewValue(0xa0c62a0, 0x82d0800, 0x8238600, 0x82d0500, 0xa1040e0, 0x8238600, 0xa105020, 0x197, 0x0, 0xa0a1340)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:661 +0x10a
github.com/goccy/go-yaml.(*Decoder).decodeStruct(0xa0c62a0, 0x8261800, 0xa105020, 0x199, 0x82d03c0, 0xa1040c0, 0x8052cd1, 0xa097e28)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:914 +0x1098
github.com/goccy/go-yaml.(*Decoder).decodeValue(0xa0c62a0, 0x8261800, 0xa105020, 0x199, 0x82d03c0, 0xa1040c0, 0x98, 0x82455a0)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:562 +0x1af4
github.com/goccy/go-yaml.(*Decoder).createDecodedNewValue(0xa0c62a0, 0x82d0800, 0x8261800, 0x82d03c0, 0xa1040c0, 0xa0af390, 0x0, 0x0, 0xa097e18, 0xf7fa6a68)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:661 +0x10a
github.com/goccy/go-yaml.(*Decoder).decodeMap(0xa0c62a0, 0x8249680, 0xa097df0, 0x195, 0x82d0400, 0xa104140, 0xa097e00, 0x71)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:1170 +0x36a
github.com/goccy/go-yaml.(*Decoder).decodeValue(0xa0c62a0, 0x8249680, 0xa097df0, 0x195, 0x82d0400, 0xa104140, 0xb, 0x28)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:550 +0x19e3
github.com/goccy/go-yaml.(*Decoder).createDecodedNewValue(0xa0c62a0, 0x82d0800, 0x8249680, 0x82d0400, 0xa104140, 0x8249680, 0xa0d6ed8, 0x195, 0x28, 0xa0a12f4)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:661 +0x10a
github.com/goccy/go-yaml.(*Decoder).decodeStruct(0xa0c62a0, 0x826c900, 0xa0d6eb0, 0x199, 0x82d03c0, 0xa099100, 0x826e160, 0x8254c20)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:914 +0x1098
github.com/goccy/go-yaml.(*Decoder).decodeValue(0xa0c62a0, 0x826c900, 0xa0d6eb0, 0x199, 0x82d03c0, 0xa099100, 0x657, 0x0)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:562 +0x1af4
github.com/goccy/go-yaml.(*Decoder).decode(0xa0c62a0, 0x82311c0, 0xa0d6eb0, 0x16, 0x82311c0, 0x0)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:1335 +0xa0
github.com/goccy/go-yaml.(*Decoder).Decode(0xa0c62a0, 0x82311c0, 0xa0d6eb0, 0xa096590, 0xa096590)
        /go/pkg/mod/github.com/goccy/go-yaml@v1.8.0/decode.go:1364 +0x1d9
github.com/fe3dback/go-arch-lint/spec/archfile.NewYamlSpec(0xa0e2900, 0x657, 0x857, 0x857, 0xa0963c0, 0x829499c)
        /tmp/builder/spec/archfile/archfile.go:57 +0x1e6
github.com/fe3dback/go-arch-lint/cmd/container.(*Container).provideArchSpec(0xa093d14, 0x8ba)
        /tmp/builder/cmd/container/spec_archfile.go:14 +0x41
github.com/fe3dback/go-arch-lint/cmd/container.(*Container).provideSpecValidator(0xa093d14, 0xa09a2d0)
        /tmp/builder/cmd/container/spec_validator.go:12 +0x26
github.com/fe3dback/go-arch-lint/cmd/container.(*Container).ProvideSpecAnnotatedValidator(0xa093d14, 0xa0bc0f0)
        /tmp/builder/cmd/container/spec_validator.go:23 +0x25
github.com/fe3dback/go-arch-lint/cmd.checkCmdProcess(0xa0bc060, 0xffa5af82, 0x4, 0xa0a2180, 0x16, 0xa0a0260, 0xb, 0xa0a21c0, 0x20, 0x0, ...)
        /tmp/builder/cmd/cmd_check_logic.go:27 +0x11d
github.com/fe3dback/go-arch-lint/cmd.cmdCheck(0xa0d42c0, 0xa0ae0c0, 0x0, 0x2)
        /tmp/builder/cmd/cmd_check.go:11 +0x7c
github.com/spf13/cobra.(*Command).execute(0xa0d42c0, 0xa0ae0b0, 0x2, 0x2, 0xa0d42c0, 0xa0ae0b0)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x1e3
github.com/spf13/cobra.(*Command).ExecuteC(0xa0d4000, 0xa0a0010, 0x82455a0, 0x82c7a00)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x297
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:880
github.com/fe3dback/go-arch-lint/cmd.Execute(0x1)
        /tmp/builder/cmd/cmd_root.go:75 +0x150
main.run(...)
        /tmp/builder/main.go:14
main.main()
        /tmp/builder/main.go:10 +0x1a

make: *** [Makefile:89: arch] Error 1

```

_Archfile:_
```yaml
# ...
deps:
  main-http:
    canUse:
      - openapi-swagger-loader
      - env
    mayDependOn:
      - version
      -
#     ^ problem with this
```

--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@ tests-functional-update-ct:
 	go test --update
 
 arch:
-	go-arch-lint check --project-path ${PWD}
+	go run main.go check --project-path ${PWD}
 
 build-dev:
 	docker build -t fe3dback/go-arch-lint:dev .

--- a/checker/checker_test.go
+++ b/checker/checker_test.go
@@ -16,8 +16,10 @@ import (
 	"github.com/fe3dback/go-arch-lint/spec"
 )
 
-const testModulePath = "github.com/fe3dback/go-arch-lint/checker/test"
-const testArchFileV1 = ".go-arch-lint-v1.yml"
+const (
+	testModulePath = "github.com/fe3dback/go-arch-lint/checker/test"
+	testArchFileV1 = ".go-arch-lint-v1.yml"
+)
 
 func makeTestProjectRoot() string {
 	_, filename, _, _ := runtime.Caller(0)
@@ -582,10 +584,13 @@ func TestChecker_Check(t *testing.T) {
 	archFilePath := root + "/" + testArchFileV1
 	sourceCode, err := ioutil.ReadFile(archFilePath)
 	if err != nil {
-		t.Fatalf("failed read archfile '%s'", archFilePath)
+		t.Fatal(fmt.Sprintf("failed read archfile '%s'", archFilePath))
 	}
 
 	yamlSpec, err := archfile.NewYamlSpec(sourceCode)
+	if err != nil {
+		t.Fatal(fmt.Sprintf("failed to make spec: %v", err))
+	}
 
 	arch, err := spec.NewArch(
 		path.NewResolver(),
@@ -594,8 +599,7 @@ func TestChecker_Check(t *testing.T) {
 		root,
 	)
 	if err != nil {
-		t.Errorf(fmt.Sprintf("failed to make arch: %v", err))
-		return
+		t.Fatal(fmt.Sprintf("failed to make arch: %v", err))
 	}
 
 	checker := NewChecker(

--- a/spec/validator/check_deps.go
+++ b/spec/validator/check_deps.go
@@ -13,21 +13,6 @@ func withCheckerDependencies(reg checkerRegistry) {
 			return reg.utils().isKnownComponent(name)
 		})
 
-		for index, componentName := range rules.MayDependOn {
-			reg.applyChecker(fmt.Sprintf("$.deps.%s.mayDependOn[%d]", name, index), func() error {
-				return reg.utils().isKnownComponent(componentName)
-			})
-		}
-
-		for index, vendorName := range rules.CanUse {
-			reg.applyChecker(
-				fmt.Sprintf("$.deps.%s.canUse[%d]", name, index),
-				func() error {
-					return reg.utils().isKnownVendor(vendorName)
-				},
-			)
-		}
-
 		if len(rules.MayDependOn) == 0 && len(rules.CanUse) == 0 {
 			reg.applyChecker(
 				fmt.Sprintf("$.deps.%s", name),

--- a/spec/validator/check_deps_components.go
+++ b/spec/validator/check_deps_components.go
@@ -1,0 +1,36 @@
+package validator
+
+import (
+	"fmt"
+)
+
+func withCheckerComponentDependencies(reg checkerRegistry) {
+	for name, rules := range reg.spec().Dependencies {
+		name := name
+		rules := rules
+
+		existComponents := make(map[string]bool)
+
+		for index, componentName := range rules.MayDependOn {
+			componentName := componentName
+
+			if _, ok := existComponents[componentName]; ok {
+				reg.applyChecker(
+					fmt.Sprintf("$.deps.%s.mayDependOn[%d]", name, index),
+					func() error {
+						return fmt.Errorf("component dublicated")
+					},
+				)
+			}
+
+			reg.applyChecker(
+				fmt.Sprintf("$.deps.%s.mayDependOn[%d]", name, index),
+				func() error {
+					return reg.utils().isKnownComponent(componentName)
+				},
+			)
+
+			existComponents[componentName] = true
+		}
+	}
+}

--- a/spec/validator/check_deps_vendors.go
+++ b/spec/validator/check_deps_vendors.go
@@ -1,0 +1,36 @@
+package validator
+
+import (
+	"fmt"
+)
+
+func withCheckerVendorDependencies(reg checkerRegistry) {
+	for name, rules := range reg.spec().Dependencies {
+		name := name
+		rules := rules
+
+		existVendors := make(map[string]bool)
+
+		for index, vendorName := range rules.CanUse {
+			vendorName := vendorName
+
+			if _, ok := existVendors[vendorName]; ok {
+				reg.applyChecker(
+					fmt.Sprintf("$.deps.%s.canUse[%d]", name, index),
+					func() error {
+						return fmt.Errorf("vendor dublicated")
+					},
+				)
+			}
+
+			reg.applyChecker(
+				fmt.Sprintf("$.deps.%s.canUse[%d]", name, index),
+				func() error {
+					return reg.utils().isKnownVendor(vendorName)
+				},
+			)
+
+			existVendors[vendorName] = true
+		}
+	}
+}

--- a/spec/validator/registry.go
+++ b/spec/validator/registry.go
@@ -25,6 +25,8 @@ func newArchFileCheckerRegistry(spec *archfile.YamlSpec, validatorUtils *validat
 		withCheckerCommonVendors,
 		withCheckerComponents,
 		withCheckerDependencies,
+		withCheckerVendorDependencies,
+		withCheckerComponentDependencies,
 		withCheckerExcludedFiles,
 		withCheckerVendors,
 		withCheckerVersion,

--- a/spec/warnparser/source_marker.go
+++ b/spec/warnparser/source_marker.go
@@ -2,7 +2,6 @@ package warnparser
 
 type (
 	sourceMarker struct {
-		valid      bool
 		sourceLine int
 		sourcePos  int
 	}

--- a/spec/warnparser/warning_annotator.go
+++ b/spec/warnparser/warning_annotator.go
@@ -7,8 +7,10 @@ import (
 	"strings"
 )
 
-const sourceMarkerLine = ">"
-const sourceMarkerPosition = "^"
+const (
+	sourceMarkerLine     = ">"
+	sourceMarkerPosition = "^"
+)
 
 type WarningSourceParser struct {
 }
@@ -18,9 +20,9 @@ func NewWarningSourceParser() *WarningSourceParser {
 }
 
 func (a WarningSourceParser) Parse(sourceText string) (line, pos int, err error) {
-	marker := parseSourceWarning(sourceText)
-	if !marker.valid {
-		return 0, 0, fmt.Errorf("failed to parse warning source text")
+	marker, err := parseSourceWarning(sourceText)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to parse warning source text: %v", err)
 	}
 
 	return marker.sourceLine, marker.sourcePos, nil
@@ -35,57 +37,51 @@ func (a WarningSourceParser) Parse(sourceText string) (line, pos int, err error)
 //	  185 |       - game_component
 //	  186 |       - game_utils
 //	  187 |
-func parseSourceWarning(sourceText string) sourceMarker {
-	notValid := sourceMarker{valid: false}
-
-	if !strings.Contains(sourceText, sourceMarkerLine) {
-		return notValid
-	}
+//
+// Another example without line mark:
+// 	   32 |        - 3rd-cobra
+// 	   33 |   cmd:    canUse:
+// 	                  ^
+// 	   36 |       - go-modfile
+// 	   37 |
+// 	   38 |   a:
+func parseSourceWarning(sourceText string) (sourceMarker, error) {
+	notValid := sourceMarker{}
 
 	if !strings.Contains(sourceText, sourceMarkerPosition) {
-		return notValid
+		return notValid, fmt.Errorf("not found position marker")
 	}
 
 	marker := sourceMarker{
-		valid:      false,
 		sourceLine: 0,
 		sourcePos:  0,
 	}
 
-	lineFound := false
 	leftOffset := 0
+	previousLine := ""
 
 	for _, sourceLine := range strings.Split(sourceText, "\n") {
-		if lineFound {
-			// in marker pos line
-			// `                ^`
-
-			marker.sourcePos = strings.Index(sourceLine, sourceMarkerPosition) - leftOffset
-			marker.valid = true
-			break
-		}
-
-		if !strings.Contains(sourceLine, sourceMarkerLine) {
+		if !strings.Contains(sourceLine, sourceMarkerPosition) {
+			previousLine = sourceLine
 			continue
 		}
 
-		// in marker line
-		// `> 183 |       - engine`
+		// found pos marker line with "^"
+		leftOffset = strings.Index(previousLine, `|`) + 1
+		marker.sourcePos = strings.Index(sourceLine, sourceMarkerPosition) - leftOffset
 
-		matches := regexp.MustCompile(`^>\s+(\d+)\s+\|`).FindStringSubmatch(sourceLine)
+		matches := regexp.MustCompile(`^>?\s+(\d+)\s+\|`).FindStringSubmatch(previousLine)
 		if len(matches) != 2 {
-			return notValid
+			return notValid, fmt.Errorf("not found line number in '%s'", previousLine)
 		}
 
 		lineNumber, err := strconv.Atoi(matches[1])
 		if err != nil {
-			return notValid
+			return notValid, fmt.Errorf("not found line number in first match")
 		}
 
 		marker.sourceLine = lineNumber
-		lineFound = true
-		leftOffset = strings.Index(sourceLine, `|`) + 1
 	}
 
-	return marker
+	return marker, nil
 }

--- a/spec/warnparser/warning_annotator_test.go
+++ b/spec/warnparser/warning_annotator_test.go
@@ -10,9 +10,10 @@ func Test_parseSourceError(t *testing.T) {
 		sourceText string
 	}
 	tests := []struct {
-		name string
-		args args
-		want sourceMarker
+		name      string
+		args      args
+		want      sourceMarker
+		wantError bool
 	}{
 		{
 			name: "valid 183",
@@ -29,10 +30,10 @@ func Test_parseSourceError(t *testing.T) {
 `,
 			},
 			want: sourceMarker{
-				valid:      true,
 				sourceLine: 183,
 				sourcePos:  9,
 			},
+			wantError: false,
 		},
 		{
 			name: "valid 1",
@@ -44,10 +45,26 @@ func Test_parseSourceError(t *testing.T) {
    4 |`,
 			},
 			want: sourceMarker{
-				valid:      true,
 				sourceLine: 1,
 				sourcePos:  10,
 			},
+			wantError: false,
+		},
+		{
+			name: "valid, no line marker",
+			args: args{
+				sourceText: `32 |        - 3rd-cobra
+   33 |   cmd:    canUse:
+   			      ^
+   36 |       - go-modfile
+   37 |
+   38 |   a:`,
+			},
+			want: sourceMarker{
+				sourceLine: 33,
+				sourcePos:  5,
+			},
+			wantError: false,
 		},
 		{
 			name: "invalid",
@@ -55,16 +72,28 @@ func Test_parseSourceError(t *testing.T) {
 				sourceText: "",
 			},
 			want: sourceMarker{
-				valid:      false,
 				sourceLine: 0,
 				sourcePos:  0,
 			},
+			wantError: true,
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := parseSourceWarning(tt.args.sourceText); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("parseSourceError() = %v, want %v", got, tt.want)
+			got, err := parseSourceWarning(tt.args.sourceText)
+
+			if err == nil && tt.wantError {
+				t.Errorf("parseSourceWarning() = expected error, but is not, got = %v", got)
+				return
+			}
+
+			if err != nil && !tt.wantError {
+				t.Errorf("parseSourceWarning() = unexpected err: %v", err)
+				return
+			}
+
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("parseSourceWarning() = %v, want %v", got, tt.want)
 			}
 		})
 	}

--- a/test/check/invalid_spec.ct
+++ b/test/check/invalid_spec.ct
@@ -1,7 +1,5 @@
 $ go-arch-lint check --project-path ${PWD}/test/check/project --arch-file invalid_spec.yml --output-color=false --> FAIL
 module: github.com/fe3dback/go-arch-lint/test/check/project
-[Archfile] path '$.deps.cmd': unknown component 'cmd' (+annotation: failed parse source warning text for path '$.deps.cmd': failed to parse warning source text)
-[Archfile] path '$.deps.main.canUse[0]': unknown vendor '3rd-cobra' (+annotation: failed parse source warning text for path '$.deps.main.canUse[0]': failed to parse warning source text)
 [Archfile] path '$.excludeFiles[0]': error parsing regexp: missing closing ): `(A-Z`:
    3 |   depOnAnyVendor: false
    4 | 
@@ -33,17 +31,42 @@ module: github.com/fe3dback/go-arch-lint/test/check/project
            ^
   25 | deps:
   26 |   main:
-[Archfile] path '$.deps.main.mayDependOn[0]': unknown component 'cmd':
+[Archfile] path '$.deps.main.mayDependOn[1]': unknown component 'cmd':
+  26 |   main:
+  27 |     mayDependOn:
+  28 |       - not_exist_too_rnd_order       - cmd
+               ^
+  30 |     canUse:
+  31 |       - 3rd-cobra-not-defined-too       - 3rd-cobra 
+[Archfile] path '$.deps.main.canUse[0]': unknown vendor '3rd-cobra-not-defined-too':
+  28 |       - not_exist_too_rnd_order       - cmd     canUse:
+  29 |       - 3rd-cobra-not-defined-too
+               ^
+  32 |       - 3rd-cobra 
+  33 |   cmd:
+[Archfile] path '$.deps.main.mayDependOn[0]': unknown component 'not_exist_too_rnd_order':
   25 |  
   26 | deps:
   27 |   main:
 > 28 |     mayDependOn:
-  29 |       - cmd
+  29 |       - not_exist_too_rnd_order
                ^
-  29 |     canUse:
-  30 |       - 3rd-cobra 
+  29 |       - cmd     canUse:
+  30 |       - 3rd-cobra-not-defined-too 
+[Archfile] path '$.deps.main.canUse[1]': unknown vendor '3rd-cobra':
+  29 |        - cmd     canUse:
+  30 |       - 3rd-cobra-not-defined-too       - 3rd-cobra
+               ^
+  34 |   cmd:    canUse:
+[Archfile] path '$.deps.cmd': unknown component 'cmd':
+  32 |        - 3rd-cobra 
+  33 |   cmd:    canUse:
+                 ^
+  36 |       - go-modfile
+  37 | 
+  38 |   a:
 [Archfile] path '$.deps.a': should have ref in 'MayDependOn' or at least one flag of ['anyProjectDeps', 'anyVendorDeps']:
-> 36 | null:
+> 38 | null:
            ^
 
 panic: cmd: arch file invalid syntax

--- a/test/check/project/invalid_spec.yml
+++ b/test/check/project/invalid_spec.yml
@@ -25,8 +25,10 @@ commonComponents:
 deps:
   main:
     mayDependOn:
+      - not_exist_too_rnd_order # not exist dep (V)
       - cmd # not exist dep (V)
     canUse:
+      - 3rd-cobra-not-defined-too # not defined vendor (V)
       - 3rd-cobra # not defined vendor (V)
 
   cmd: # not exist component

